velox/type/Type.h

```diff
@@ -507,6 +542,96 @@ class TypeBase : public Type {
   }
 };
 
+/*
+ * This class represents the Decimal Type to represent the fixed-point numbers
+ * in Velox. The parameter "precision" represents the number of digits the
+ * Decimal Type can support and scale represents the number of digits to the
+ * right of the decimal point.
+ */
+template <TypeKind KIND>
+class DecimalType : public Type {
+ public:
+  static constexpr uint8_t kMaxPrecision =
+      KIND == TypeKind::SHORT_DECIMAL ? 18 : 38;
+  static constexpr uint8_t kMaxScale =
+      KIND == TypeKind::SHORT_DECIMAL ? 18 : 38;
+  static_assert(
+      KIND == TypeKind::SHORT_DECIMAL || KIND == TypeKind::LONG_DECIMAL);
+
+  DecimalType(const uint8_t precision = 18, const uint8_t scale = 0)
+      : Type(KIND), precision_(precision), scale_(scale) {
+    VELOX_CHECK_LE(scale, precision);
+    VELOX_CHECK_LE(precision, kMaxPrecision);
+  }
+
+  inline bool isPrimitiveType() const override {
+    return true;
+  }
+
+  inline bool isFixedWidth() const override {
+    return true;
+  }
+
+  inline const char* kindName() const override {
+    return TypeTraits<KIND>::name;
+  }
+
+  inline uint32_t size() const override {
+    // DECIMAL TypeSignature can have 2 children which can be
+    // long literals or variables to represent the precision and scale.
+    return NUM_DECIMAL_PARAMETERS;
+  }
+
+  inline bool operator==(const Type& other) const override {
+    // Comparing the string forms of decimal type should cover
+    // the TypeKind, precision and scale.
+    return this->toString().compare(other.toString()) == 0;
+  }
+
+  bool compareParameter(const int pos, std::string parameter) const {
+    if (pos == 0) {
+      std::string precision = std::to_string(precision_);
+      return precision.compare(parameter) == 0;
+    } else if (pos == 1) {
+      std::string scale = std::to_string(scale_);
+      return scale.compare(parameter) == 0;
+    }
+    throw std::invalid_argument{"Invalid parameter position"};
+  };
+
+  inline const uint8_t precision() const {
+    return precision_;
+  }
+
+  inline const uint8_t scale() const {
+    return scale_;
+  }
+
+  inline const std::shared_ptr<const Type>& childAt(uint32_t) const override {
```

Comment:
Would it be possible to use ScalarType<KIND> as a base class?

Reply:
Since I have templatized DecimalType now, inheriting from ScalarType seemed to work without any issue.